# NLTK Chapter 2 

In [19]:
import nltk
from nltk.corpus import gutenberg

In [20]:
nltk.corpus.gutenberg.fileids() #Finding text of gutenberg project

['austen-emma.txt',
 'austen-persuasion.txt',
 'austen-sense.txt',
 'bible-kjv.txt',
 'blake-poems.txt',
 'bryant-stories.txt',
 'burgess-busterbrown.txt',
 'carroll-alice.txt',
 'chesterton-ball.txt',
 'chesterton-brown.txt',
 'chesterton-thursday.txt',
 'edgeworth-parents.txt',
 'melville-moby_dick.txt',
 'milton-paradise.txt',
 'shakespeare-caesar.txt',
 'shakespeare-hamlet.txt',
 'shakespeare-macbeth.txt',
 'whitman-leaves.txt']

In [250]:
emma=nltk.corpus.gutenberg.words('austen-emma.txt')
len(emma)

192427

In [251]:
nltk.Text(emma).concordance('surprize') #You need to convert it to nltk.Text

Displaying 25 of 37 matches:
er father , was sometimes taken by surprize at his being still able to pity ` 
hem do the other any good ." " You surprize me ! Emma must do Harriet good : a
Knightley actually looked red with surprize and displeasure , as he stood up ,
r . Elton , and found to his great surprize , that Mr . Elton was actually on 
d aid ." Emma saw Mrs . Weston ' s surprize , and felt that it must be great ,
father was quite taken up with the surprize of so sudden a journey , and his f
y , in all the favouring warmth of surprize and conjecture . She was , moreove
he appeared , to have her share of surprize , introduction , and pleasure . Th
ir plans ; and it was an agreeable surprize to her , therefore , to perceive t
talking aunt had taken me quite by surprize , it must have been the death of m
f all the dialogue which ensued of surprize , and inquiry , and congratulation
 the present . They might chuse to surprize her ." Mrs . Cole had many to agre
the mode of it , the my

In [23]:
from nltk.corpus import gutenberg

In [24]:
for file in gutenberg.fileids():
    num_char=len(gutenberg.raw(file))#No charachters
    num_words=len(gutenberg.words(file))#NUmber of words
    num_sents=len(gutenberg.sents(file)) #Number of sentences
    num_vocab=len(set([w.lower() for w in gutenberg.words(file)])) #unique words
    print((int(num_char/num_words), int(num_words/num_sents), int(num_words/num_vocab), file))#Average world length, Average sentence length
    #How frequenctly a word is used (lexical Diversity Score)
    #The average world length is 4. But it's actually 3. Why? Because it the code counts the spaces too.

(4, 24, 26, 'austen-emma.txt')
(4, 26, 16, 'austen-persuasion.txt')
(4, 28, 22, 'austen-sense.txt')
(4, 33, 79, 'bible-kjv.txt')
(4, 19, 5, 'blake-poems.txt')
(4, 19, 14, 'bryant-stories.txt')
(4, 17, 12, 'burgess-busterbrown.txt')
(4, 20, 12, 'carroll-alice.txt')
(4, 20, 11, 'chesterton-ball.txt')
(4, 22, 11, 'chesterton-brown.txt')
(4, 18, 10, 'chesterton-thursday.txt')
(4, 20, 24, 'edgeworth-parents.txt')
(4, 25, 15, 'melville-moby_dick.txt')
(4, 52, 10, 'milton-paradise.txt')
(4, 11, 8, 'shakespeare-caesar.txt')
(4, 12, 7, 'shakespeare-hamlet.txt')
(4, 12, 6, 'shakespeare-macbeth.txt')
(4, 36, 12, 'whitman-leaves.txt')


### Time for Assessing Web and Chat Text (Less formal text)

In [25]:
from nltk.corpus import webtext

In [26]:
for field in webtext.fileids():
    print(field, webtext.raw(field)[:65])

firefox.txt Cookie Manager: "Don't allow sites that set removed cookies to se
grail.txt SCENE 1: [wind] [clop clop clop] 
KING ARTHUR: Whoa there!  [clop
overheard.txt White guy: So, do you have any plans for this evening?
Asian girl
pirates.txt PIRATES OF THE CARRIBEAN: DEAD MAN'S CHEST, by Ted Elliott & Terr
singles.txt 25 SEXY MALE, seeks attrac older single lady, for discreet encoun
wine.txt Lovely delicate, fragrant Rhone wine. Polished leather and strawb


### Brown Corpus

In [27]:
from nltk.corpus import brown
brown.categories()

['adventure',
 'belles_lettres',
 'editorial',
 'fiction',
 'government',
 'hobbies',
 'humor',
 'learned',
 'lore',
 'mystery',
 'news',
 'religion',
 'reviews',
 'romance',
 'science_fiction']

In [28]:
news_words=brown.words(categories='news')
news_words

['The', 'Fulton', 'County', 'Grand', 'Jury', 'said', ...]

In [29]:
len(news_words)

100554

In [30]:
w=brown.words(categories='fiction') #Words from a category 'ficiton'
w

['Thirty-three', 'Scotty', 'did', 'not', 'go', 'back', ...]

In [31]:
len(w)

68488

In [32]:
brown.words(fileids=['cg22']) #Words from a file cg22

['Does', 'our', 'society', 'have', 'a', 'runaway', ',', ...]

In [33]:
words_multiple=brown.words(categories=['fiction','news','editorial']) 
words_multiple

['The', 'Fulton', 'County', 'Grand', 'Jury', 'said', ...]

In [196]:
len(words_multiple)

230646

######  Brown corpus can be used to learn systematic differences between different genres.

In [197]:
news_text=brown.words(categories='news')
fdist= nltk.FreqDist(w.lower() for w in news_text)
modals = ['can', 'could', 'may', 'might', 'must', 'will']
for m in modals:
    print((m, fdist[m]))

('can', 94)
('could', 87)
('may', 93)
('might', 38)
('must', 53)
('will', 389)


In [ ]:
type(fdist)

In [205]:
cfd = nltk.ConditionalFreqDist((genre, word) for genre in brown.categories() for word in brown.words(categories=genre))

In [204]:
genres = ['news', 'religion', 'hobbies', 'science_fiction', 'romance', 'humor']
modal = ['can', 'could', 'may', 'might', 'must', 'will']
cfd.tabulate(conditions=genres,  samples=modals)

                  can could   may might  must  will 
           news    93    86    66    38    50   389 
       religion    82    59    78    12    54    71 
        hobbies   268    58   131    22    83   264 
science_fiction    16    49     4    12     8    16 
        romance    74   193    11    51    45    43 
          humor    16    30     8     8     9    13 


In [214]:
genres = ['news', 'religion', 'hobbies', 'science_fiction', 'romance', 'humor']
genre_words=[(genre, word)for genre in genres for word in brown.words(categories=genre) if len(word)>20]
print(genre_words)

[('news', 'Republican-controlled'), ('news', 'Universal-International'), ('news', 'Television-Electronics'), ('news', 'Scotch-Irish-Scandinavian'), ('news', 'Washington-Alexandria'), ('news', 'heavy-electrical-goods'), ('news', 'heavy-electrical-goods'), ('news', 'collective-bargaining'), ('news', 'advertising-conscious'), ('religion', 'culture-Protestantism'), ('religion', "Preparation-Inquirers'"), ('religion', 'never-to-be-forgotten'), ('hobbies', 'chest-back-lat-shoulder'), ('hobbies', 'definition-specialization'), ('hobbies', 'definition-specialization'), ('hobbies', 'contraction-extension'), ('hobbies', 'vertical-takeoff-and-landing'), ('hobbies', 'Schubert-Beethoven-Mozart'), ('hobbies', 'mailed-fist-in-velvet-glove'), ('hobbies', "let's-make-your-house-our-club"), ('hobbies', 'composer-pianist-conductor'), ('hobbies', 'Commission-controlled'), ('hobbies', 'individual-contributor'), ('hobbies', 'individual-contributor'), ('hobbies', 'engineering-management'), ('hobbies', 'engine

In [215]:
cfd=nltk.ConditionalFreqDist(genre_words)
cfd

ConditionalFreqDist(nltk.probability.FreqDist,
                    {'hobbies': FreqDist({'Commission-controlled': 1,
                               'Schubert-Beethoven-Mozart': 1,
                               'build-better-for-less': 1,
                               'chest-back-lat-shoulder': 1,
                               'composer-pianist-conductor': 1,
                               'contraction-extension': 1,
                               'definition-specialization': 2,
                               'engineering-management': 2,
                               'individual-contributor': 2,
                               "let's-make-your-house-our-club": 1,
                               'mailed-fist-in-velvet-glove': 1,
                               'organization-position': 1,
                               'three-men-and-a-helper': 1,
                               'too-simple-to-be-true': 1,
                               'vertical-takeoff-and-landing': 1}),
               

######  Word counts can distinguish genres.

In [234]:
genres=['hobbies','science_fiction', 'romance']
word_length=[(genre, len(word))for genre in genres for word in brown.words(categories=genre) if len(word)>16]
print(word_length)

[('hobbies', 19), ('hobbies', 23), ('hobbies', 19), ('hobbies', 25), ('hobbies', 25), ('hobbies', 21), ('hobbies', 28), ('hobbies', 25), ('hobbies', 20), ('hobbies', 19), ('hobbies', 17), ('hobbies', 18), ('hobbies', 17), ('hobbies', 18), ('hobbies', 18), ('hobbies', 17), ('hobbies', 19), ('hobbies', 18), ('hobbies', 19), ('hobbies', 19), ('hobbies', 19), ('hobbies', 18), ('hobbies', 20), ('hobbies', 17), ('hobbies', 27), ('hobbies', 30), ('hobbies', 19), ('hobbies', 26), ('hobbies', 17), ('hobbies', 17), ('hobbies', 19), ('hobbies', 19), ('hobbies', 18), ('hobbies', 18), ('hobbies', 17), ('hobbies', 17), ('hobbies', 18), ('hobbies', 18), ('hobbies', 18), ('hobbies', 18), ('hobbies', 19), ('hobbies', 19), ('hobbies', 19), ('hobbies', 17), ('hobbies', 18), ('hobbies', 18), ('hobbies', 20), ('hobbies', 17), ('hobbies', 21), ('hobbies', 18), ('hobbies', 22), ('hobbies', 22), ('hobbies', 22), ('hobbies', 22), ('hobbies', 21), ('hobbies', 19), ('hobbies', 18), ('hobbies', 18), ('hobbies', 1

In [237]:
cfd=nltk.ConditionalFreqDist(word_length)
cfd.tabulate(conditions=genres, samples=[i for i in range(17, 21)])

                17 18 19 20 
        hobbies 19 17 15  3 
science_fiction  1  2  2  0 
        romance  5  2  4  1 


In [262]:
news_text=brown.words(categories='hobbies')
print(nltk.Text(news_text).concordance('swim'))

Displaying 3 of 3 matches:
ions to `` come over next summer and swim in our new pool '' were both unexpec
eing our lingo for those who come to swim and sink into our bar while protesti
e when neighborhood teen-agers might swim at definite hours . This has saved u
None


In [263]:
news_text=brown.words(categories='hobbies')
print(nltk.Text(news_text).collocations())

New York; chemical name; United States; Drug's chemical; per head;
years ago; interior design; Santa Cruz; milligrams per; nuclear
weapons; locking bars; St. Sophia; combustion chamber; bacterial
diarrhea; per day; locking bar; Junior Showmanship; grams per; air
conditioning; per pound
None


###  Reutuers Corpus (news documents)

###### Categorised into 90 topics and grouped into two sets 'Training' and 'Testing'.

In [ ]:
from nltk.corpus import reuters
reuters.fileids()

In [ ]:
reuters.categories() #These categories overlap with each other.

In [ ]:
reuters.categories('training/9865')

In [ ]:
reuters.categories(['training/9865', 'training/9866', 'training/9880'])

In [ ]:
reuters.fileids('barley')

In [ ]:
reuters.fileids(['barley', 'corn'])

In [ ]:
reuters.words(categories='barley')

In [ ]:
reuters.words(categories=['barley','corn'])

In [ ]:
reuters.words(fileids=['training/9865', 'training/9880'])

In [ ]:
reuters.words(['training/9865', 'training/9880'])

###  Inaugural Address Corpus (of Presidents?)

In [ ]:
from nltk.corpus import inaugural
inaugural.fileids()

In [ ]:
[files[:4] for files in inaugural.fileids()]

In [ ]:
cfd=nltk.ConditionalFreqDist((target, file[:4]) for file in inaugural.fileids() for w in inaugural.words(file)
                             for target in ['america','citizen'] if w.lower().startswith(target)) 
cfd.plot()

In [ ]:
cfd

### Conditional Frequency

######  FreqDist returns a list of frequencies for each word. Conditional Frequency returns a list of Frequency Distribution and each frequency distribution has a condition associated with it. For example if the condition is genre meaning for each genre get its frequency distribution. 

In [ ]:
g=[(genre, word) for genre in ['news', 'romance'] for word in brown.words(categories=genre)]
print(g[:4])
print(g[-4:])

In [ ]:
cfd= nltk.ConditionalFreqDist((genre, word) for genre in ['news', 'romance'] for word in brown.words(categories=genre))
cfd.conditions() #It has two conditions news and romance

In [ ]:
cfd= nltk.ConditionalFreqDist((genre, word) for genre in ['news', 'romance','adventure'] for word in brown.words(categories=genre))
cfd.conditions()

###  Plotting and Tabulating Distributions

######  In the plot() and tabulate() we can exploit the conditions parameter and specify which condtions to display. By default we get all the conditions.

In [198]:
from nltk.corpus import udhr
languages = ['English', 'Chickasaw', 'German_Deutsch', 'Greenlandic_Inuktikut', 'Hungarian_Magyar', 'Ibibio_Efik']
cfd = nltk.ConditionalFreqDist((lang, len(word)) for lang in languages for word in udhr.words(lang+'-Latin1'))
cfd.tabulate(conditions=['English', 'German_Deutsch'], samples=range(10), cumulative=True) #samples takes a list of keys to use

                  0    1    2    3    4    5    6    7    8    9 
       English    0  185  525  883  997 1166 1283 1440 1558 1638 
German_Deutsch    0  171  263  614  717  894 1013 1110 1213 1275 


###### We interpret above example first row last cell that there are 1638 words with length 9 or lower.

In [199]:
cfd

ConditionalFreqDist(nltk.probability.FreqDist,
                    {'Chickasaw': FreqDist({1: 411,
                               2: 99,
                               3: 41,
                               4: 68,
                               5: 91,
                               6: 89,
                               7: 77,
                               8: 70,
                               9: 49,
                               10: 33,
                               11: 16,
                               12: 28,
                               13: 45,
                               14: 10,
                               15: 6,
                               16: 4,
                               17: 5,
                               18: 3,
                               19: 2,
                               20: 1,
                               21: 1,
                               23: 1}),
                     'English': FreqDist({1: 185,
                               2: 340,
       

In [200]:
days=['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
brown.categories()

['adventure',
 'belles_lettres',
 'editorial',
 'fiction',
 'government',
 'hobbies',
 'humor',
 'learned',
 'lore',
 'mystery',
 'news',
 'religion',
 'reviews',
 'romance',
 'science_fiction']

In [201]:
cfd=nltk.ConditionalFreqDist((genre, word) for genre in ['news', 'romance'] for word in brown.words(categories=genre) if word in days)
cfd.tabulate(sample=days, cumulative=False)

           Friday    Monday  Saturday    Sunday  Thursday   Tuesday Wednesday 
   news        41        54        33        51        20        43        22 
romance         3         2         4         5         1         3         3 


### Generating random text with Bigrams

###### bigrams() takes a list of words and returns a list of consecutive word pairs

In [ ]:
def generate_model(cfdist, word, num=15):
    for i in range(num):
        print(word)
        word=cfdist[word].max()

In [ ]:
text=nltk.corpus.genesis.words('english-kjv.txt')
bigrams=nltk.bigrams(text)
cfd=nltk.ConditionalFreqDist(bigrams)
cfd['living']

In [ ]:
generate_model(cfd, 'living')

In [ ]:
cfd['creature']

### Wordlist Corpora

In [ ]:
def unusual_words(text):
    text_vocab=set(w.lower() for w in text if w.isalpha())
    english_vocab=set(w.lower() for w in nltk.corpus.words.words())
    unusual_words=text_vocab.difference(english_vocab)
    return sorted(unusual_words)

In [ ]:
unusual_words(gutenberg.words('austen-sense.txt'))

######  Stopwords are words like or, the, a, an wtc which don't have a lot of lexical content.

In [ ]:
from nltk.corpus import stopwords
sorted(stopwords.words('english'))

In [ ]:
def content_fraction(text):
    """returns fraction of words not in the stopwords"""
    stopwords=nltk.corpus.stopwords.words('english')
    content=[w for w in text if w.lower() not in stopwords]
    return len(content)/len(text)

In [ ]:
content_fraction(reuters.words())

###### Above we have used a lexical resource to filter the contents of text resource. 2 corpusses.

In [ ]:
puzzle_letters= nltk.FreqDist('egivrconl')
obligatory='r'
wordlist=nltk.corpus.words.words()
[w for w in wordlist if len(w)>= 6 and obligatory in w and nltk.FreqDist(w) <= puzzle_letters]

In [ ]:
puzzle_letters

In [ ]:
from nltk.corpus import names
cfd=nltk.ConditionalFreqDist((file,name[-1]) for file in names.fileids() for name in names.words(file))
cfd.plot()

######  Female names end with a.

### Pronouncing Dictionary

###### CMU Pronouncing Dictionary for English is designed for speech synthesis. Richer lexical resource because with a word it also has addtional properties

In [37]:
entries=nltk.corpus.cmudict.entries()
len(entries)

133737

In [38]:
entries[3043:3051] #The second list elements are called phones which are distinct labels for each contrastive sound of the word

[('aloisio', ['AA0', 'L', 'OY1', 'S', 'IY0', 'OW0']),
 ('aloka', ['AH0', 'L', 'OW1', 'K', 'AH0']),
 ('alomar', ['AE1', 'L', 'AH0', 'M', 'AA2', 'R']),
 ('alon', ['AH0', 'L', 'AA1', 'N']),
 ('alone', ['AH0', 'L', 'OW1', 'N']),
 ('along', ['AH0', 'L', 'AO1', 'NG']),
 ('alonge', ['AE1', 'L', 'AH0', 'N', 'JH']),
 ('alongi', ['AA0', 'L', 'OW1', 'NG', 'G', 'IY0'])]

In [39]:
for word, pron in entries:
    if len(pron)==3:
        ph1, ph2, ph3 = pron
        if ph1=='P' and ph3=='T':
            print((word, ph2))

('pait', 'EY1')
('pat', 'AE1')
('pate', 'EY1')
('patt', 'AE1')
('peart', 'ER1')
('peat', 'IY1')
('peet', 'IY1')
('peete', 'IY1')
('pert', 'ER1')
('pet', 'EH1')
('pete', 'IY1')
('pett', 'EH1')
('piet', 'IY1')
('piette', 'IY1')
('pit', 'IH1')
('pitt', 'IH1')
('pot', 'AA1')
('pote', 'OW1')
('pott', 'AA1')
('pout', 'AW1')
('puett', 'UW1')
('purt', 'ER1')
('put', 'UH1')
('putt', 'AH1')


######  The next program finds words whose pronunciation ends with syllable sounding like nicks

In [40]:
syllable=['N', 'IH0','K','S']
[w for w, pron in entries if pron[-4:] == syllable]

["atlantic's",
 'audiotronics',
 'avionics',
 'beatniks',
 'calisthenics',
 'centronics',
 'chamonix',
 'chetniks',
 "clinic's",
 'clinics',
 'conics',
 'conics',
 'cryogenics',
 'cynics',
 'diasonics',
 "dominic's",
 'ebonics',
 'electronics',
 "electronics'",
 "endotronics'",
 'endotronics',
 'enix',
 'environics',
 'ethnics',
 'eugenics',
 'fibronics',
 'flextronics',
 'harmonics',
 'hispanics',
 'histrionics',
 'identics',
 'ionics',
 'kibbutzniks',
 'lasersonics',
 'lumonics',
 'mannix',
 'mechanics',
 "mechanics'",
 'microelectronics',
 'minix',
 'minnix',
 'mnemonics',
 'mnemonics',
 'molonicks',
 'mullenix',
 'mullenix',
 'mullinix',
 'mulnix',
 "munich's",
 'nucleonics',
 'onyx',
 'organics',
 "panic's",
 'panics',
 'penix',
 'pennix',
 'personics',
 'phenix',
 "philharmonic's",
 'phoenix',
 'phonics',
 'photronics',
 'pinnix',
 'plantronics',
 'pyrotechnics',
 'refuseniks',
 "resnick's",
 'respironics',
 'sconnix',
 'siliconix',
 'skolniks',
 'sonics',
 'sputniks',
 'technics

### Mismatches between pronunciation and writing

In [41]:
[w for w, pros in entries if pros[-1].lower()=='m' and w[-1]=='n'] 

['autumn', 'column', 'condemn', 'damn', 'goddamn', 'hymn', 'solemn']

In [42]:
sorted(set([w[:2] for w, pros in entries if pros[0].lower() == 'n' and w[0] != 'n']))

['gn', 'kn', 'mn', 'pn']

###  The phones contain digits to contain primary stress(1), secondary stress(2), and not stress(0)

In [43]:
def stress(pros):
    return [w for phone in pros for w in phone if w.isdigit()]


In [44]:
[w for w, pros in entries if stress(pros)==['0','1','0','2', '0']]

['abbreviated',
 'abbreviated',
 'abbreviating',
 'accelerated',
 'accelerating',
 'accelerator',
 'accelerators',
 'accentuated',
 'accentuating',
 'accommodated',
 'accommodating',
 'accommodative',
 'accumulated',
 'accumulating',
 'accumulative',
 'accumulator',
 'accumulators',
 'accusatory',
 'adenovirus',
 'adjudicated',
 'adjudicating',
 'administrating',
 'administrative',
 'administrator',
 "administrators'",
 "administrator's",
 'administrators',
 'adulterated',
 'adventurism',
 'adventurism',
 'affiliated',
 'affiliated',
 "affiliated's",
 'affiliating',
 'alleviated',
 'alleviated',
 'alleviating',
 'alliteration',
 'alliterative',
 'amalgamated',
 "amalgamated's",
 'amalgamating',
 'ameliorated',
 'ameridata',
 'amoxicillin',
 'anachronism',
 'anachronisms',
 'annihilated',
 'annihilating',
 'antagonism',
 'antagonisms',
 'antagonizing',
 'anticipated',
 'anticipated',
 'anticipating',
 'apologizes',
 'apologizing',
 'apothecary',
 'appreciated',
 'appreciating',
 'apprec

In [46]:
p3 = [(pron[0]+'-'+pron[2], word) for (word, pron) in entries if pron[0]=='P' and len(pron)==3]
cfd=nltk.ConditionalFreqDist(p3)
for template in cfd.conditions():
    if len(cfd[template])>10:
        words=cfd[template].keys()
        wordlist=" ".join(words)
        print(template, wordlist[:70]+"..")

P-P paap paape pap pape papp paup peep pep pip pipe pipp poop pop pope pop..
P-R paar pair par pare parr pear peer pier poor poore por pore porr pour..
P-K pac pack paek paik pak pake paque peak peake pech peck peek perc perk ..
P-S pace pass pasts peace pearse pease perce pers perse pesce piece piss p..
P-L pahl pail paille pal pale pall paul paule paull peal peale pearl pearl..
P-N paign pain paine pan pane pawn payne peine pen penh penn pin pine pinn..
P-Z pais paiz pao's pas pause paws pays paz peas pease pei's perz pez pies..
P-T pait pat pate patt peart peat peet peete pert pet pete pett piet piett..
P-CH patch pautsch peach perch petsch petsche piche piech pietsch pitch pit..
P-UW1 peru peugh pew plew plue prew pru prue prugh pshew pugh..


In [51]:
sorted(p3)

[('P-AA1', 'pla'),
 ('P-AH0', 'pera'),
 ('P-AH0', 'pia'),
 ('P-AW1', 'plough'),
 ('P-AW1', 'plow'),
 ('P-AW1', 'prough'),
 ('P-AW1', 'prow'),
 ('P-AY1', 'ply'),
 ('P-AY1', 'pri'),
 ('P-AY1', 'pry'),
 ('P-B', 'pub'),
 ('P-CH', 'patch'),
 ('P-CH', 'pautsch'),
 ('P-CH', 'peach'),
 ('P-CH', 'perch'),
 ('P-CH', 'petsch'),
 ('P-CH', 'petsche'),
 ('P-CH', 'piche'),
 ('P-CH', 'piech'),
 ('P-CH', 'pietsch'),
 ('P-CH', 'pitch'),
 ('P-CH', 'pitsch'),
 ('P-CH', 'poach'),
 ('P-CH', 'poche'),
 ('P-CH', 'pooch'),
 ('P-CH', 'pouch'),
 ('P-CH', 'puche'),
 ('P-CH', 'putsch'),
 ('P-D', 'pad'),
 ('P-D', 'paid'),
 ('P-D', 'pawed'),
 ('P-D', 'peed'),
 ('P-D', 'pied'),
 ('P-D', 'pod'),
 ('P-D', 'poohed'),
 ('P-ER0', 'payer'),
 ('P-ER0', 'poer'),
 ('P-ER0', 'power'),
 ('P-ER0', 'poyer'),
 ('P-ER0', 'pyre'),
 ('P-ER1', 'payeur'),
 ('P-EY1', 'play'),
 ('P-EY1', 'pray'),
 ('P-EY1', 'prey'),
 ('P-F', 'paff'),
 ('P-F', 'poff'),
 ('P-F', 'poof'),
 ('P-F', 'puff'),
 ('P-G', 'peg'),
 ('P-G', 'pegg'),
 ('P-G', 'pig'),

In [52]:
cfd

ConditionalFreqDist(nltk.probability.FreqDist,
                    {'P-AA1': FreqDist({'pla': 1}),
                     'P-AH0': FreqDist({'pera': 1, 'pia': 1}),
                     'P-AW1': FreqDist({'plough': 1,
                               'plow': 1,
                               'prough': 1,
                               'prow': 1}),
                     'P-AY1': FreqDist({'ply': 1, 'pri': 1, 'pry': 1}),
                     'P-B': FreqDist({'pub': 1}),
                     'P-CH': FreqDist({'patch': 1,
                               'pautsch': 1,
                               'peach': 1,
                               'perch': 1,
                               'petsch': 1,
                               'petsche': 1,
                               'piche': 1,
                               'piech': 1,
                               'pietsch': 1,
                               'pitch': 1,
                               'pitsch': 1,
                               'poach': 1,
 

In [54]:
prondict = nltk.corpus.cmudict.dict()
prondict['fire']

[['F', 'AY1', 'ER0'], ['F', 'AY1', 'R']]

In [60]:
print(prondict)

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.


In [59]:
text = ['natural', 'language', 'processing']
print([ph for w in text for ph in prondict[w][0]])

['N', 'AE1', 'CH', 'ER0', 'AH0', 'L', 'L', 'AE1', 'NG', 'G', 'W', 'AH0', 'JH', 'P', 'R', 'AA1', 'S', 'EH0', 'S', 'IH0', 'NG']


In [63]:
prondict['natural'][0]

['N', 'AE1', 'CH', 'ER0', 'AH0', 'L']

### Comparative Wordlists

In [65]:
from nltk.corpus import swadesh
print(swadesh.fileids())

['be', 'bg', 'bs', 'ca', 'cs', 'cu', 'de', 'en', 'es', 'fr', 'hr', 'it', 'la', 'mk', 'nl', 'pl', 'pt', 'ro', 'ru', 'sk', 'sl', 'sr', 'sw', 'uk']


In [67]:
print(swadesh.words('en'))

['I', 'you (singular), thou', 'he', 'we', 'you (plural)', 'they', 'this', 'that', 'here', 'there', 'who', 'what', 'where', 'when', 'how', 'not', 'all', 'many', 'some', 'few', 'other', 'one', 'two', 'three', 'four', 'five', 'big', 'long', 'wide', 'thick', 'heavy', 'small', 'short', 'narrow', 'thin', 'woman', 'man (adult male)', 'man (human being)', 'child', 'wife', 'husband', 'mother', 'father', 'animal', 'fish', 'bird', 'dog', 'louse', 'snake', 'worm', 'tree', 'forest', 'stick', 'fruit', 'seed', 'leaf', 'root', 'bark (from tree)', 'flower', 'grass', 'rope', 'skin', 'meat', 'blood', 'bone', 'fat (noun)', 'egg', 'horn', 'tail', 'feather', 'hair', 'head', 'ear', 'eye', 'nose', 'mouth', 'tooth', 'tongue', 'fingernail', 'foot', 'leg', 'knee', 'hand', 'wing', 'belly', 'guts', 'neck', 'back', 'breast', 'heart', 'liver', 'drink', 'eat', 'bite', 'suck', 'spit', 'vomit', 'blow', 'breathe', 'laugh', 'see', 'hear', 'know (a fact)', 'think', 'smell', 'fear', 'sleep', 'live', 'die', 'kill', 'fight',

###### We can find cognate (similar) words in multiple languages using entries() method.

In [69]:
fr2en=swadesh.entries(['fr','en'])#French and English
print(fr2en)                       

[('I', 'je'), ('you (singular), thou', 'tu, vous'), ('he', 'il'), ('we', 'nous'), ('you (plural)', 'vous'), ('they', 'ils, elles'), ('this', 'ceci'), ('that', 'cela'), ('here', 'ici'), ('there', 'là'), ('who', 'qui'), ('what', 'quoi'), ('where', 'où'), ('when', 'quand'), ('how', 'comment'), ('not', 'ne...pas'), ('all', 'tout'), ('many', 'plusieurs'), ('some', 'quelques'), ('few', 'peu'), ('other', 'autre'), ('one', 'un'), ('two', 'deux'), ('three', 'trois'), ('four', 'quatre'), ('five', 'cinq'), ('big', 'grand'), ('long', 'long'), ('wide', 'large'), ('thick', 'épais'), ('heavy', 'lourd'), ('small', 'petit'), ('short', 'court'), ('narrow', 'étroit'), ('thin', 'mince'), ('woman', 'femme'), ('man (adult male)', 'homme'), ('man (human being)', 'homme'), ('child', 'enfant'), ('wife', 'femme, épouse'), ('husband', 'mari, époux'), ('mother', 'mère'), ('father', 'père'), ('animal', 'animal'), ('fish', 'poisson'), ('bird', 'oiseau'), ('dog', 'chien'), ('louse', 'pou'), ('snake', 'serpent'), ('w

In [72]:
translate = dict(fr2en)
translate['name']

'nom'

## WordNet

Semantically oriented dictionary of English.

### Senses and Synonyms

In [73]:
from nltk.corpus import wordnet as wn
wn.synsets('motorcar')

[Synset('car.n.01')]

The above answer means that the word 'motorcar' has just one possible meaning, the forst noun set of car synset means a synonymset(collection of synonyms). 

In [75]:
wn.synset('car.n.01').lemma_names()

['car', 'auto', 'automobile', 'machine', 'motorcar']

Each word of synset can have a different meaning. However we're intersted in the common meaning in the above synset.

In [78]:
wn.synset('car.n.01').definition()

'a motor vehicle with four wheels; usually propelled by an internal combustion engine'

In [80]:
wn.synset('car.n.01').examples()

['he needs a car to get to work']

To eliminate ambiguity we will identify these words as car.n.01.automobile and car.n.01.motorcar and so on.

In [88]:
wn.synset('car.n.01').lemmas() #The pairing of a synset is lemma

[Lemma('car.n.01.car'),
 Lemma('car.n.01.auto'),
 Lemma('car.n.01.automobile'),
 Lemma('car.n.01.machine'),
 Lemma('car.n.01.motorcar')]

In [89]:
wn.lemma('car.n.01.automobile').name()

'automobile'

In [90]:
wn.lemma('car.n.01.automobile').synset()

Synset('car.n.01')

In [91]:
wn.synsets('car') #The word car is ambiguous 

[Synset('car.n.01'),
 Synset('car.n.02'),
 Synset('car.n.03'),
 Synset('car.n.04'),
 Synset('cable_car.n.01')]

In [92]:
wn.synsets('motorcar') #The word motorcar is unambiguous 

[Synset('car.n.01')]

In [99]:
for syn in wn.synsets('car'):
    print(syn.lemma_names())

['car', 'auto', 'automobile', 'machine', 'motorcar']
['car', 'railcar', 'railway_car', 'railroad_car']
['car', 'gondola']
['car', 'elevator_car']
['cable_car', 'car']


In [100]:
wn.lemmas('car') #We can access all the lemmas having word 'car' using this method

[Lemma('car.n.01.car'),
 Lemma('car.n.02.car'),
 Lemma('car.n.03.car'),
 Lemma('car.n.04.car'),
 Lemma('cable_car.n.01.car')]

In [101]:
wn.synsets('dish')

[Synset('dish.n.01'),
 Synset('dish.n.02'),
 Synset('dish.n.03'),
 Synset('smasher.n.02'),
 Synset('dish.n.05'),
 Synset('cup_of_tea.n.01'),
 Synset('serve.v.06'),
 Synset('dish.v.02')]

In [103]:
for w in wn.synsets('dish'):
    print(w.lemma_names())

['dish']
['dish']
['dish', 'dishful']
['smasher', 'stunner', 'knockout', 'beauty', 'ravisher', 'sweetheart', 'peach', 'lulu', 'looker', 'mantrap', 'dish']
['dish', 'dish_aerial', 'dish_antenna', 'saucer']
['cup_of_tea', 'bag', 'dish']
['serve', 'serve_up', 'dish_out', 'dish_up', 'dish']
['dish']


In [106]:
wn.lemmas('dish')

[Lemma('dish.n.01.dish'),
 Lemma('dish.n.02.dish'),
 Lemma('dish.n.03.dish'),
 Lemma('smasher.n.02.dish'),
 Lemma('dish.n.05.dish'),
 Lemma('cup_of_tea.n.01.dish'),
 Lemma('serve.v.06.dish'),
 Lemma('dish.v.02.dish')]

In [266]:
for syn in wn.synsets('dish'):
    print(syn.lemma_names(), ' : ' + syn.definition())

['dish']  : a piece of dishware normally used as a container for holding or serving food
['dish']  : a particular item of prepared food
['dish', 'dishful']  : the quantity that a dish will hold
['smasher', 'stunner', 'knockout', 'beauty', 'ravisher', 'sweetheart', 'peach', 'lulu', 'looker', 'mantrap', 'dish']  : a very attractive or seductive looking woman
['dish', 'dish_aerial', 'dish_antenna', 'saucer']  : directional antenna consisting of a parabolic reflector for microwave or radio frequency radiation
['cup_of_tea', 'bag', 'dish']  : an activity that you like or at which you are superior
['serve', 'serve_up', 'dish_out', 'dish_up', 'dish']  : provide (usually but not necessarily food)
['dish']  : make concave; shape like a dish


In [267]:
for syn in wn.synsets('train'):
    print(syn.lemma_names(), ' : ' + syn.definition())

['train', 'railroad_train']  : public transport provided by a line of railway cars coupled together and drawn by a locomotive
['string', 'train']  : a sequentially ordered set of things or events or ideas in which each successive member is related to the preceding
['caravan', 'train', 'wagon_train']  : a procession (of wagons or mules or camels) traveling together in single file
['train']  : a series of consequences wrought by an event
['train']  : piece of cloth forming the long back section of a gown that is drawn along the floor
['gearing', 'gear', 'geartrain', 'power_train', 'train']  : wheelwork consisting of a connected set of rotating gears by which force is transmitted or motion or torque is changed
['train', 'develop', 'prepare', 'educate']  : create by training and teaching
['train', 'prepare']  : undergo training or instruction in preparation for a particular role, function, or profession
['discipline', 'train', 'check', 'condition']  : develop (children's) behavior by instr

In [109]:
motorcar=wn.synset('car.n.01')

In [268]:
wn.synsets('prepare')

[Synset('fix.v.12'),
 Synset('cook.v.02'),
 Synset('prepare.v.03'),
 Synset('organize.v.05'),
 Synset('prepare.v.05'),
 Synset('train.v.01'),
 Synset('prepare.v.07'),
 Synset('train.v.02')]

In [269]:
wn.synsets('cook')

[Synset('cook.n.01'),
 Synset('cook.n.02'),
 Synset('cook.v.01'),
 Synset('cook.v.02'),
 Synset('cook.v.03'),
 Synset('fudge.v.01'),
 Synset('cook.v.05')]

In [274]:
wn.synsets('misrepresent')

[Synset('misrepresent.v.01'), Synset('fudge.v.01')]

In [270]:
for syn in wn.synsets('cook'):
    print(syn.lemma_names(), ' : ' + syn.definition())

['cook']  : someone who cooks food
['Cook', 'James_Cook', 'Captain_Cook', 'Captain_James_Cook']  : English navigator who claimed the east coast of Australia for Britain and discovered several Pacific islands (1728-1779)
['cook']  : prepare a hot meal
['cook', 'fix', 'ready', 'make', 'prepare']  : prepare for eating by applying heat
['cook']  : transform and make suitable for consumption by heating
['fudge', 'manipulate', 'fake', 'falsify', 'cook', 'wangle', 'misrepresent']  : tamper, with the purpose of deception
['cook']  : transform by heating


In [275]:
motorcar=wn.synset('car.n.01')
types_of_motorcar = motorcar.hyponyms()
types_of_motorcar[26]

Synset('stanley_steamer.n.01')

In [292]:
print(sorted([lemma.name() for syn in types_of_motorcar for lemma in syn.lemmas()]))

['Model_T', 'S.U.V.', 'SUV', 'Stanley_Steamer', 'ambulance', 'beach_waggon', 'beach_wagon', 'bus', 'cab', 'compact', 'compact_car', 'convertible', 'coupe', 'cruiser', 'electric', 'electric_automobile', 'electric_car', 'estate_car', 'gas_guzzler', 'hack', 'hardtop', 'hatchback', 'heap', 'horseless_carriage', 'hot-rod', 'hot_rod', 'jalopy', 'jeep', 'landrover', 'limo', 'limousine', 'loaner', 'minicar', 'minivan', 'pace_car', 'patrol_car', 'phaeton', 'police_car', 'police_cruiser', 'prowl_car', 'race_car', 'racer', 'racing_car', 'roadster', 'runabout', 'saloon', 'secondhand_car', 'sedan', 'sport_car', 'sport_utility', 'sport_utility_vehicle', 'sports_car', 'squad_car', 'station_waggon', 'station_wagon', 'stock_car', 'subcompact', 'subcompact_car', 'taxi', 'taxicab', 'tourer', 'touring_car', 'two-seater', 'used-car', 'waggon', 'wagon']


In [279]:
wn.synset('stanley_steamer.n.01').hypernyms()

[Synset('car.n.01')]

In [278]:
wn.synset('car.n.01').hypernyms()

[Synset('motor_vehicle.n.01')]

In [118]:
paths=motorcar.hypernym_paths()
len(paths)

2

In [119]:
paths

[[Synset('entity.n.01'),
  Synset('physical_entity.n.01'),
  Synset('object.n.01'),
  Synset('whole.n.02'),
  Synset('artifact.n.01'),
  Synset('instrumentality.n.03'),
  Synset('container.n.01'),
  Synset('wheeled_vehicle.n.01'),
  Synset('self-propelled_vehicle.n.01'),
  Synset('motor_vehicle.n.01'),
  Synset('car.n.01')],
 [Synset('entity.n.01'),
  Synset('physical_entity.n.01'),
  Synset('object.n.01'),
  Synset('whole.n.02'),
  Synset('artifact.n.01'),
  Synset('instrumentality.n.03'),
  Synset('conveyance.n.03'),
  Synset('vehicle.n.01'),
  Synset('wheeled_vehicle.n.01'),
  Synset('self-propelled_vehicle.n.01'),
  Synset('motor_vehicle.n.01'),
  Synset('car.n.01')]]

In [121]:
[synset.name() for synset in paths[0]]

['entity.n.01',
 'physical_entity.n.01',
 'object.n.01',
 'whole.n.02',
 'artifact.n.01',
 'instrumentality.n.03',
 'container.n.01',
 'wheeled_vehicle.n.01',
 'self-propelled_vehicle.n.01',
 'motor_vehicle.n.01',
 'car.n.01']

Hypernyms and hyponyms are called lexical relations because they relate one synset
to another. These two relations navigate up and down the “is-a” hierarchy. Another
important way to navigate the WordNet network is from items to their components
(meronyms) or to the things they are contained in (holonyms). For example, the parts
of a tree are its trunk, crown, and so on; these are the part_meronyms(). The substance
a tree is made of includes heartwood and sapwood, i.e., the substance_meronyms(). A
collection of trees forms a forest, i.e., the member_holonyms():

In [122]:
wn.synsets('tree')

[Synset('tree.n.01'),
 Synset('tree.n.02'),
 Synset('tree.n.03'),
 Synset('corner.v.02'),
 Synset('tree.v.02'),
 Synset('tree.v.03'),
 Synset('tree.v.04')]

In [123]:
wn.synset('tree.n.01').part_meronyms()

[Synset('burl.n.02'),
 Synset('crown.n.07'),
 Synset('limb.n.02'),
 Synset('stump.n.01'),
 Synset('trunk.n.01')]

In [124]:
wn.synset('tree.n.01').substance_meronyms()

[Synset('heartwood.n.01'), Synset('sapwood.n.01')]

In [125]:
wn.synset('tree.n.01').member_holonyms()

[Synset('forest.n.01')]

In [133]:
wn.synsets('mint')

[Synset('batch.n.02'),
 Synset('mint.n.02'),
 Synset('mint.n.03'),
 Synset('mint.n.04'),
 Synset('mint.n.05'),
 Synset('mint.n.06'),
 Synset('mint.v.01'),
 Synset('mint.s.01')]

In [129]:
wn.synsets("mint", wn.NOUN) #return only synset where it is a noun.

[Synset('batch.n.02'),
 Synset('mint.n.02'),
 Synset('mint.n.03'),
 Synset('mint.n.04'),
 Synset('mint.n.05'),
 Synset('mint.n.06')]

In [130]:
wn.synset('mint.n.02').definition() #Here it is a noun

'any of the forms of Chinese spoken in Fukien province'

In [134]:
wn.synset('mint.v.01').definition()#Here it is not a noun henc why it was removed

'form by stamping, punching, or printing'

In [140]:
for synset in wn.synsets('mint', wn.NOUN):
    print(synset.name()+ ' : ' + synset.definition())

batch.n.02 : (often followed by `of') a large number or amount or extent
mint.n.02 : any north temperate plant of the genus Mentha with aromatic leaves and small mauve flowers
mint.n.03 : any member of the mint family of plants
mint.n.04 : the leaves of a mint plant used fresh or candied
mint.n.05 : a candy that is flavored with a mint oil
mint.n.06 : a plant where money is coined by authority of the government


In [144]:
print(wn.synset('batch.n.02').lemma_names())

['batch', 'deal', 'flock', 'good_deal', 'great_deal', 'hatful', 'heap', 'lot', 'mass', 'mess', 'mickle', 'mint', 'mountain', 'muckle', 'passel', 'peck', 'pile', 'plenty', 'pot', 'quite_a_little', 'raft', 'sight', 'slew', 'spate', 'stack', 'tidy_sum', 'wad']


In [145]:
wn.synset('mint.n.04').part_holonyms()

[Synset('mint.n.02')]

In [148]:
wn.synset('mint.n.04').substance_holonyms()

[Synset('mint.n.05')]

There are also relationships between verbs. For example, the act of walking involves
the act of stepping, so walking entails stepping. Some verbs have multiple entailments

In [149]:
wn.synset('walk.v.01').entailments()

[Synset('step.v.01')]

In [150]:
wn.synset('eat.v.01').entailments()

[Synset('chew.v.01'), Synset('swallow.v.01')]

In [152]:
wn.synset('tease.v.03').entailments()

[Synset('arouse.v.07'), Synset('disappoint.v.01')]

In [153]:
wn.synset('tease.v.03').definition()

'to arouse hope, desire, or curiosity without satisfying them'

In [157]:
for syn in wn.synsets('tease'):
    print(syn.lemma_names(), syn.definition())

['tease', 'teaser', 'annoyer', 'vexer'] someone given to teasing (as by mocking or stirring curiosity)
['coquette', 'flirt', 'vamp', 'vamper', 'minx', 'tease', 'prickteaser'] a seductive woman who uses her sex appeal to exploit men
['tease', 'teasing', 'ribbing', 'tantalization'] the act of harassing someone playfully or maliciously (especially by ridicule); provoking someone with persistent annoyances
['tease', 'badger', 'pester', 'bug', 'beleaguer'] annoy persistently
['tease', 'razz', 'rag', 'cod', 'tantalize', 'tantalise', 'bait', 'taunt', 'twit', 'rally', 'ride'] harass with persistent criticism or carping
['tease'] to arouse hope, desire, or curiosity without satisfying them
['tease'] tear into pieces
['tease'] raise the nap of (fabrics)
['tease', 'tease_apart', 'loosen'] disentangle and raise the fibers of
['tease', 'card'] separate the fibers of
['tease'] mock or make fun of playfully
['tease', 'fluff'] ruffle (one's hair) by combing the ends towards the scalp, for a full effec

###### Antonyms

In [160]:
wn.lemma('supply.n.02.supply').antonyms()

[Lemma('demand.n.02.demand')]

In [161]:
wn.lemma('rush.v.01.rush').antonyms()

[Lemma('linger.v.04.linger')]

In [162]:
wn.lemma('horizontal.a.01.horizontal').antonyms()

[Lemma('inclined.a.02.inclined'), Lemma('vertical.a.01.vertical')]

In [166]:
wn.lemma('staccato.r.01.staccato').antonyms()

[Lemma('legato.r.01.legato')]

In [171]:
#to view the lexical relations
print(dir(wn.synset('harmony.n.02')))

['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__slots__', '__str__', '__subclasshook__', '__unicode__', '__weakref__', '_all_hypernyms', '_definition', '_examples', '_frame_ids', '_hypernyms', '_instance_hypernyms', '_iter_hypernym_lists', '_lemma_names', '_lemma_pointers', '_lemmas', '_lexname', '_max_depth', '_min_depth', '_name', '_needs_root', '_offset', '_pointers', '_pos', '_related', '_shortest_hypernym_paths', '_wordnet_corpus_reader', 'also_sees', 'attributes', 'causes', 'closure', 'common_hypernyms', 'definition', 'entailments', 'examples', 'frame_ids', 'hypernym_distances', 'hypernym_paths', 'hypernyms', 'hyponyms', 'instance_hypernyms', 'instance_hyponyms', 'jcn_similarity', 'lch_similarity', 'lemma_names', 'lemmas', 'lexnam

In [172]:
wn.synsets('create')

[Synset('make.v.03'),
 Synset('create.v.02'),
 Synset('create.v.03'),
 Synset('create.v.04'),
 Synset('create.v.05'),
 Synset('produce.v.02')]

In [178]:
for syn in wn.synsets('create'):
    print(syn.lemma_names(), ' : ' + syn.definition())

['make', 'create']  : make or cause to be or to become
['create']  : bring into existence
['create']  : pursue a creative activity; be engaged in a creative activity
['create']  : invest with a new title, office, or rank
['create', 'make']  : create by artistic means
['produce', 'make', 'create']  : create or manufacture a man-made product


In [182]:
wn.synset('make.v.03').lemmas()

[Lemma('make.v.03.make'), Lemma('make.v.03.create')]

Recall that each synset has one or more hypernym paths that link it to a root hypernym
such as entity.n.01. Two synsets linked to the same root may have several hypernyms
in common. If two synsets share a very specific hypernym—one that
is low down in the hypernym hierarchy—they must be closely related.

In [294]:
right = wn.synset('right_whale.n.01')
orca = wn.synset('orca.n.01')
minke = wn.synset('minke_whale.n.01')
tortoise = wn.synset('tortoise.n.01')
novel = wn.synset('novel.n.01')
right.lowest_common_hypernyms(orca)

[Synset('whale.n.02')]

In [185]:
right.lowest_common_hypernyms(minke)

[Synset('baleen_whale.n.01')]

In [186]:
right.lowest_common_hypernyms(tortoise)

[Synset('vertebrate.n.01')]

In [187]:
right.lowest_common_hypernyms(novel)

[Synset('entity.n.01')]

Of course we know that whale is very specific (and baleen whale even more so), whereas
vertebrate is more general and entity is completely general. We can quantify this concept
of generality by looking up the depth of each synset

In [189]:
wn.synset('whale.n.02').min_depth()

13

In [295]:
wn.synset('vertebrate.n.01').min_depth()

8

In [190]:
wn.synset('baleen_whale.n.01').min_depth()

14

In [191]:
wn.synset('entity.n.01').min_depth()

0

Similarity measures have been defined over the collection of WordNet synsets that
incorporate this insight. For example, path_similarity assigns a score in the range
0–1 based on the shortest path that connects the concepts in the hypernym hierarchy
(-1 is returned in those cases where a path cannot be found). Comparing a synset with
itself will return 1. Consider the following similarity scores, relating right whale to minke
whale, orca, tortoise, and novel. Although the numbers won’t mean much, they decrease
as we move away from the semantic space of sea creatures to inanimate objects.

In [192]:
right.path_similarity(minke)

0.25

In [193]:
right.path_similarity(orca)

0.16666666666666666

In [195]:
right.path_similarity(novel)

0.043478260869565216

###### NLTK also includes VerbNet

In [217]:
simple_list=[('category 1',"hello"), ('category 1',"hi"), ('category 1',"hello"), ('category 0',"hello"),
             ('category 0',"hi"),('category 1',"Bye")]
cfd=nltk.ConditionalFreqDist(simple_list)
cfd

ConditionalFreqDist(nltk.probability.FreqDist,
                    {'category 0': FreqDist({'hello': 1, 'hi': 1}),
                     'category 1': FreqDist({'Bye': 1, 'hello': 2, 'hi': 1})})

In [283]:
from nltk.corpus import wordnet as wn
wn.synsets('misrepresent')

[Synset('misrepresent.v.01'), Synset('fudge.v.01')]

In [287]:
wn.synset('misrepresent.v.01').definition()

'represent falsely'

In [288]:
wn.synset('misrepresent.v.01').lemmas()

[Lemma('misrepresent.v.01.misrepresent'), Lemma('misrepresent.v.01.belie')]